File: src/BmSDK.Host/runtime.cpp
#include <cassert>
#include <filesystem>
#include <iostream>

#include <coreclr_delegates.h>
#include <hostfxr.h>
#include <nethost.h>
#include <Windows.h>

#include "runtime.h"

using namespace std;
using namespace std::filesystem;

namespace Runtime {
    // Func aliases for hostfxr
    using HostInitFn = hostfxr_initialize_for_runtime_config_fn;
    using HostGetDelegateFn = hostfxr_get_runtime_delegate_fn;
    using HostCloseFn = hostfxr_close_fn;
    using HostLoadAssemblyFn = load_assembly_and_get_function_pointer_fn;
    // Entry point func type alias
    typedef void(CORECLR_DELEGATE_CALLTYPE* ComponentEntryFn)();

    // Funcs from hostfxr
    static HostInitFn hostInit = nullptr;
    static HostGetDelegateFn hostGetDelegate = nullptr;
    static HostCloseFn hostClose = nullptr;
    static HostLoadAssemblyFn hostLoadAssembly = nullptr;

    // Filesystem helpers
    static path GetPluginPath() {
        wstring result(MAX_PATH, L'\0');
        GetModuleFileName(NULL, result.data(), MAX_PATH);
        return path(result);
    }

    static path GetGameDir() { return GetPluginPath().parent_path(); }

    static void* LoadLibFromPath(const char_t* path) {
        HMODULE h = LoadLibrary(path);
        assert(h != nullptr);
        return (void*)h;
    }

    static void* GetExport(void* h, const char* name) {
        void* f = GetProcAddress((HMODULE)h, name);
        assert(f != nullptr);
        return f;
    }

    // Using the nethost library, discover the location of hostfxr and get exports
    static bool LoadHostFxr() {
        // Pre-allocate a large buffer for the path to hostfxr
        char_t buffer[MAX_PATH];
        size_t bufferSize = sizeof(buffer) / sizeof(char_t);
        int rc = get_hostfxr_path(buffer, &bufferSize, nullptr);
        if (rc != 0)
            return false;

        // Load hostfxr and get desired exports
        void* lib = LoadLibFromPath(buffer);
        hostInit = (HostInitFn)GetExport(lib, "hostfxr_initialize_for_runtime_config");
        hostGetDelegate = (HostGetDelegateFn)GetExport(lib, "hostfxr_get_runtime_delegate");
        hostClose = (HostCloseFn)GetExport(lib, "hostfxr_close");

        return (hostInit && hostGetDelegate && hostClose);
    }

    // Load and initialize .NET Core and get desired function pointer for scenario
    static HostLoadAssemblyFn GetDotnetLoadAssembly(const char_t* config_path) {
        // Load .NET Core
        void* hostLoadAssembly = nullptr;
        hostfxr_handle cxt = nullptr;
        int rc = hostInit(config_path, nullptr, &cxt);
        if (rc != 0 || cxt == nullptr) {
            std::cerr << "Init failed: " << std::hex << std::showbase << rc << std::endl;
            hostClose(cxt);
            return nullptr;
        }

        // Get the load assembly function pointer
        rc = hostGetDelegate(
            cxt,
            hdt_load_assembly_and_get_function_pointer,
            &hostLoadAssembly);
        if (rc != 0 || hostLoadAssembly == nullptr)
            std::cerr << "Get delegate failed: " << std::hex << std::showbase << rc << std::endl;

        hostClose(cxt);
        return (HostLoadAssemblyFn)hostLoadAssembly;
    }

    void LoadDll() {
        // Load HostFxr and get exported hosting functions
        if (!LoadHostFxr()) {
            assert(false && "Failure: LoadHostFxr()");
            return;
        }

        // Initialize and start the .NET Core runtime
        const wstring basePath = GetGameDir();
        const wstring asmPath = L"\\sdk\\BmSDK";
        const wstring configPath = basePath + asmPath + L".runtimeconfig.json";
        hostLoadAssembly = GetDotnetLoadAssembly(configPath.c_str());
        assert(hostLoadAssembly != nullptr && "Failure: GetDotnetLoadAssembly()");

        // Load managed assembly and get function pointer to a managed method
        const wstring dotnetDllPath = basePath + asmPath + L".dll";
        const wstring dotnetType = L"BmSDK.Framework.Loader, BmSDK";
        const wstring dotnetMethod = L"GuardedDllMain";

        // Function pointer to managed delegate
        ComponentEntryFn GuardedDllMain = nullptr;
        int rc = hostLoadAssembly(
            dotnetDllPath.c_str(),
            dotnetType.c_str(),
            dotnetMethod.c_str(),
            UNMANAGEDCALLERSONLY_METHOD,
            nullptr,
            (void**)&GuardedDllMain);

        if (rc != 0 || GuardedDllMain == nullptr) {
            assert(false && "Failure: hostLoadAssembly()");
            return;
        }

        GuardedDllMain();
    }
}

File: src/BmSDK.Host/Runtime.cs
using System.Runtime.InteropServices;
using Windows.Win32;
using Windows.Win32.Foundation;
using static BmSDK.Host.HostfxrNative;

namespace BmSDK.Host;

static class Runtime
{
    static unsafe string GetGameDir()
    {
        Span<char> buffer = stackalloc char[260]; // MAX_PATH
        fixed (char* bufPtr = buffer)
        {
            PInvoke.GetModuleFileName(HMODULE.Null, bufPtr, (uint)buffer.Length);
        }

        return Path.GetDirectoryName(buffer.TrimEnd('\0').ToString())!;
    }

    static bool LoadHostfxr()
    {
        // Use nethost to discover hostfxr location
        Span<char> buffer = stackalloc char[260];
        int bufferSize = buffer.Length;
        int rc = get_hostfxr_path(buffer, ref bufferSize, 0);
        if (rc != 0)
            return false;

        // Load the hostfxr library so P/Invoke can resolve it
        string hostfxrPath = buffer[..bufferSize].TrimEnd('\0').ToString();
        PInvoke.LoadLibrary(hostfxrPath);

        return true;
    }

    static LoadAssemblyAndGetFunctionPointerFn? GetDotnetLoadAssembly(string configPath)
    {
        int rc = hostfxr_initialize_for_runtime_config(configPath, 0, out nint cxt);
        if (rc != 0 || cxt == 0)
        {
            Console.Error.WriteLine($"Init failed: 0x{rc:X}");
            if (cxt != 0) hostfxr_close(cxt);
            return null;
        }

        rc = hostfxr_get_runtime_delegate(
            cxt,
            HostfxrDelegateType.hdt_load_assembly_and_get_function_pointer,
            out nint loadAssemblyPtr);

        hostfxr_close(cxt);

        if (rc != 0 || loadAssemblyPtr == 0)
        {
            Console.Error.WriteLine($"Get delegate failed: 0x{rc:X}");
            return null;
        }

        return Marshal.GetDelegateForFunctionPointer<LoadAssemblyAndGetFunctionPointerFn>(loadAssemblyPtr);
    }

    public static void LoadDll()
    {
        // Load hostfxr
        if (!LoadHostfxr())
        {
            Console.Error.WriteLine("Failure: LoadHostfxr()");
            return;
        }

        // Initialize .NET Core runtime
        string gameDir = GetGameDir();
        string configPath = Path.Combine(gameDir, "sdk", "BmSDK.runtimeconfig.json");
        var loadAssembly = GetDotnetLoadAssembly(configPath);
        if (loadAssembly is null)
        {
            Console.Error.WriteLine("Failure: GetDotnetLoadAssembly()");
            return;
        }

        // Load managed assembly and get entry point
        string dllPath = Path.Combine(gameDir, "sdk", "BmSDK.dll");
        int rc = loadAssembly(
            dllPath,
            "BmSDK.Framework.Loader, BmSDK",
            "GuardedDllMain",
            UNMANAGEDCALLERSONLY_METHOD,
            0,
            out nint entryPoint);

        if (rc != 0 || entryPoint == 0)
        {
            Console.Error.WriteLine($"Failure: LoadAssemblyAndGetFunctionPointer() 0x{rc:X}");
            return;
        }

        // Call GuardedDllMain
        var guardedDllMain = Marshal.GetDelegateForFunctionPointer<Action>(entryPoint);
        guardedDllMain();
    }
}
